src/sgame/lua/Entity.cpp

```diff
@@ -0,0 +1,230 @@
+along with Unvanquished Source Code.  If not, see <http://www.gnu.org/licenses/>.
+
+In addition, the Unvanquished Source Code is also subject to certain additional terms.
+You should have received a copy of these additional terms immediately following the
+terms and conditions of the GNU General Public License which accompanied the Unvanquished
+Source Code.  If not, please request a copy in writing from id Software at the address
+below.
+
+If you have questions concerning this license or the applicable additional terms, you
+may contact in writing id Software LLC, c/o ZeniMax Media Inc., Suite 120, Rockville,
+Maryland 20850 USA.
+
+===========================================================================
+*/
+
+#include "sgame/lua/Entity.h"
+
+#include "sgame/lua/EntityProxy.h"
+#include "shared/lua/LuaLib.h"
+#include "shared/lua/Utils.h"
+#include "sgame/sg_local.h"
+
+using Shared::Lua::LuaLib;
+using Shared::Lua::RegType;
+
+/// Handle interactions with Entities.
+// @module entity
+
+namespace Lua {
+// These are freed by the lua GC.
+std::vector<EntityProxy*> Entity::proxies( MAX_GENTITIES );
+
+EntityProxy* Entity::CreateProxy( gentity_t* ent, lua_State* L )
+{
+	int entNum = ent->num();
+	if ( !Entity::proxies[ entNum ] || !Entity::proxies[ entNum ]->ent )
+	{
+		Entity::proxies[ entNum ] = new EntityProxy( ent, L );
+	}
+	return Entity::proxies[ entNum ];
+}
+
+namespace {
+
+/// FindById an entity based on their id. This id must be set manually prior.
+// @function find_by_id
+// @tparam string id The entity id.
+// @treturn EntityProxy|nil Returns EntityProxy if it finds a match or nil.
+// @within entity
+int FindById( lua_State* L )
+{
+	const char* id = luaL_checkstring( L, 1 );
+	gentity_t* ent = nullptr;
+	while ( ( ent = G_IterateEntities( ent ) ) )
+	{
+		if ( ent->id && !strncmp( id, ent->id, strlen( ent->id ) ) )
+		{
+			EntityProxy* proxy = Entity::CreateProxy( ent, L );
+			LuaLib<EntityProxy>::push( L, proxy );
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+/// Allows iterating over a group of entities given a class name.
+// @function iterate_classname
+// @tparam string class_name The class name to search for.
+// @treturn EntityProxy...|nil Returns one or more EntityProxies if any matches or nil.
+// @within entity
+int IterateByClassName( lua_State* L )
+{
+	const char* name = luaL_checkstring( L, 1 );
+	gentity_t* ent = nullptr;
+	int ret = 0;
+	while ( ( ent = G_IterateEntities( ent ) ) )
+	{
+		if ( !Q_stricmp( ent->classname, name ) )
+		{
+			EntityProxy* proxy = Entity::CreateProxy( ent, L );
+			LuaLib<EntityProxy>::push( L, proxy );
+			ret++;
+		}
+	}
+	return ret;
+}
+
+/// Fetch an entity by its entity number via entity[idx]. Do not use index() directly.
+// @function index
+// @tparam integer index The entity number.
+// @treturn EntityProxy|nil Returns EntityProxy if there one exists or nil.
+// @usage local ent = sgame.entity[0] -- Get the first entity.
+// @within entity
+int index( lua_State* L )
```

Comment:
This doesn't work well for players. I don't want to have to refetch a player every time.

Reply:
Then use a client number or `Client` object in the cases where you want to keep the reference beyond death.

Reply:
That is a worse developer experience. Client num means you have to fetch it every time. A client has no reference back to the parent entity.

Reply:
The way I imagine it should work is that an API that accepts an entity argument should be able to take any of an entity object, a client object, or an entity number.

Your Juggernaut implementation is a good example of something that would benefit from GentityRef semantics. Currently it has a bug that if a player suicides, it continues being the juggernaut instead of choosing a different one. If you stored the current juggernaut as a GentityRef, then you could easily detect any weird condition causing the juggernaut to disappear by checking whether the ref is valid. You wouldn't have to worry about hooking all of the possibilities like disconnection, team change, or suicide.

Reply:
> The way I imagine it should work is that an API that accepts an entity argument should be able to take any of an entity object, a client object, or an entity number.

That's why too confusing imo. Just pick one and don't overload.

> Currently it has a bug that if a player suicides, it continues being the juggernaut instead of choosing a different one.

The die logic has a bug that doesn't handle this. Note that fetching the entity every time does not fix this.

> You wouldn't have to worry about hooking all of the possibilities like disconnection, team change, or suicide.

You still need to do this. Just because you need to fetch the entity every time does not mean you can ignore the fact that a disconnection/team change/etc occurred. After all, you're storing the entityNum now since storing an entity is bad because it could be invalidated at any time.

Why not just make the proxy return nil when inuse is false instead of every generation change? It sounds like that's the best of both worlds.

Reply:
Also note that in my initial attempt, I did try to use a gentityref, it was just super unergonomic so I had to revert it.

https://github.com/Unvanquished/Unvanquished/commit/02a89eddc3a2ce609893da5bdf5d71dff87c056b

Reply:
My proposal is to have a GentityRef pointing to the current juggernaut and check every frame if it is empty, in which case a new one needs to be randomly selected. This would fix the suicide bug because the reference is invalidated a second or two after the player dies.

For scoreboard tracking, a stable client identifier would be used, same as now. Team change/disconnect hooks would still be needed, but only for clearing the scoreboard data of the player who left. I don't see when you would ever need to fetch the entity for a client number.

> Why not just make the proxy return nil when inuse is false instead of every generation change? It sounds like that's the best of both worlds.

This is not an adequate solution for detecting entity expiration, even if you check every frame. For example if bot fill is active, and someone switches teams, it often happens that a bot client is removed and a new one immediately created with the same entity number.

Reply:
I don't think this works well in practice. The entire point of Lua is to avoid being a per-frame thing. We should be working on events (die, team change, player spawn, etc). Having to refresh a counter every time is just extra annoyance.